lib/route/link/bridge.c

```diff
@@ -675,6 +910,20 @@ static int bridge_compare(struct rtnl_link *_a, struct rtnl_link *_b,
 	diff |= _DIFF(BRIDGE_ATTR_HWMODE, a->b_hwmode != b->b_hwmode);
 	diff |= _DIFF(BRIDGE_ATTR_CONFIG_MODE, a->b_config_mode != b->b_config_mode);
 
+#define _MST_HASH(msti, state) \
+	({ \
+		uint64_t val = (uint64_t)(state) << 16 | (uint64_t)(msti + 1); \
+		val ^= val << 13; val ^= val >> 7; val ^= val << 17; val; })
+
+	nl_list_for_each_entry(entry, &a->mst_list, list_node) {
+		hash ^= _MST_HASH(entry->msti, entry->state);
+	}
+	nl_list_for_each_entry(entry, &b->mst_list, list_node) {
+		hash ^= _MST_HASH(entry->msti, entry->state);
+	}
```

Comment:
This doesn't seem right. With `^` we indeed do some hashing of the values, but we want a equality implementation that bails out if there are any differences

Reply:
This should be equivalent to using two different variables for each hash and comparing them at the end because of the way XOR works. This relies on each list not having duplicate items which is true.

```c
	nl_list_for_each_entry(entry, &a->mst_list, list_node) {
		hash_1 ^= _MST_HASH(entry->msti, entry->state);
	}
	nl_list_for_each_entry(entry, &b->mst_list, list_node) {
		hash_2 ^= _MST_HASH(entry->msti, entry->state);
	}
#undef _MST_HASH
	diff |= _DIFF(BRIDGE_ATTR_MST, hash_1 != hash_2);
```

I'm kind of hesitant about the hash method altogether, since there's a small chance of false positives or false negatives. I could use two nested for loops or keep the lists sorted so comparison is easier.

Reply:
my point was that the equality check via the hash can have false negatives (wrongly not detecting inequality).

I don't think you need to do a order-independent comparison. While in some sense such configurations may be equivalent, if the order differs, we can assume that the overall object is different. That simplifies the implementation, and I probably would expect to consider ordering in the comparison.